File: Leetcode/Number of Islands/NumberOfIslands.java
/*Here we use Depth First Search (DFS) to solve this problem*/
class Solution 
{
    public int numIslands(char[][] grid) 
    {
        int count = 0;                          //declaration and initialization
        for(int i=0; i< grid.length; i++)       //to loop through the given 2D-grid of characters
        {
            for(int j=0; j<grid[i].length; j++)
            {
                if(grid[i][j]== '1')            //if we find a land
                {
                    count += 1;                 //increment counter variable
                    callDFS(grid,i,j);          //function calling
                }
            }
        }
        return count;				//return required number of islands
    }

    /*user-defined method to find the structure of the whole island by searching horizontally and
    vertically for all connected '1's (using DFS) */

    public void callDFS(char[][] grid, int i, int j)
    {
	//boundary check (so that we remain within the grid)
        if(i<0 || i>=grid.length || j<0 || j>=grid[i].length || grid[i][j]=='0')
            return;
        
        grid[i][j] = '0';       //change the value of visited land to 0
        
        callDFS(grid,i+1,j);    //check down (next row, same column)
        callDFS(grid,i-1,j);    //check up   (previous row, same column)
        callDFS(grid,i,j-1);    //check left (same row, previous column)
        callDFS(grid,i,j+1);    //check right(same row, next column)
    }
}

File: Leetcode/Number of islands/Number of Islands.py
"""
A python program to implement Number of Islands
Given an m x n 2d grid map of '1's (land) and '0's (water), return the number of islands.
An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

"""

class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        # BFS with marks
        if grid is None or len(grid) == 0:
            return 0
        islands = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '1':
                    self.explore(grid, i, j)
                    islands += 1
        return islands

    def explore(self, grid, i, j):
        grid[i][j] = 'X'
        if i - 1 >= 0 and grid[i - 1][j] == '1':
            self.explore(grid, i - 1, j)
        if j - 1 >= 0 and grid[i][j - 1] == '1':
            self.explore(grid, i, j - 1)
        if i + 1 < len(grid) and grid[i + 1][j] == '1':
            self.explore(grid, i + 1, j)
        if j + 1 < len(grid[i]) and grid[i][j + 1] == '1':
            self.explore(grid, i, j + 1)
            
"""
Sample Input/Output:
Example 1:

Input: grid = [
  ["1","1","1","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","0","0"]
]
Output: 1

Example 2:
Input: grid = [
  ["1","1","0","0","0"],
  ["1","1","0","0","0"],
  ["0","0","1","0","0"],
  ["0","0","0","1","1"]
]
Output: 3

"""
